torch/lib/c10d/TCPStore.cpp

```diff
@@ -516,6 +552,15 @@ bool TCPStore::deleteKey(const std::string& key) {
   return (numDeleted == 1);
 }
 
+void TCPStore::watchKey(const std::string& key, CallbackFunction callback) {
+  std::string regKey = regularPrefix_ + key;
+
+  tcpStoreListener_->addCallback(regKey, callback);
```

Comment:
Would I be correct if I assume the thread-safety is based on the assumption that callback function itself does not call any TCPStore APIs? For example, if I register a callback that will register another callback when fired, will it still be thread-safe? Probably not because they are called in two different threads (main thread and the `tcpStoreListener_` thread)? It's totally fine to keep it so for simplicity for now, as we fully control the callback implementation in the foreseeable future. But let's highlight that risk in the `addCallback`/`CallbackFunction` API comments.